crates/biome_js_analyze/src/semantic_analyzers/nursery/use_for_of.rs

```diff
@@ -0,0 +1,384 @@
+    ///    console.log(i, array[i]);
+    ///  }
+    /// ```
+    ///
+    /// ```js
+    /// for (let i = 1; i < array.length; i++) {
+    ///    console.log(i, array[i]);
+    ///  }
+    /// ```
+    ///
+    pub(crate) UseForOf {
+        version: "1.4.0",
+        name: "useForOf",
+        recommended: false,
+    }
+}
+
+declare_node_union! {
+    pub(crate) AnyIncrementableLike = JsPostUpdateExpression | JsPreUpdateExpression | JsAssignmentExpression
+}
+
+impl Rule for UseForOf {
+    type Query = Semantic<JsForStatement>;
+    type State = ();
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let model = ctx.model();
+        let initializer = node.initializer()?;
+
+        if !is_initializer_valid(&initializer)? {
+            return None;
+        }
+
+        let declarators = initializer.as_js_variable_declaration()?.declarators();
+        let initializer = declarators.first()?.ok()?;
+        let initializer_id = initializer.id().ok()?;
+        let test = node.test()?;
+        let binding = initializer_id
+            .as_any_js_binding()?
+            .as_js_identifier_binding()?;
+
+        if !is_test_valid(binding, &test)? {
+            return None;
+        }
+
+        if !is_update_valid(binding, node.update()?)? {
+            return None;
+        }
+
+        let body = node.body().ok()?;
+        let block = body.as_js_block_statement();
+        let references = get_initializer_references(block, binding, model);
+        let array_right = test.as_js_binary_expression()?.right().ok()?;
+        let array_used_in_for = array_right
+            .as_js_static_member_expression()?
+            .object()
+            .ok()?;
+
+        let index_only_used_with_array = |reference| {
+            let array_in_use = reference_being_used_by_array(reference, &array_used_in_for)
+                .is_some_and(|array_in_use| array_in_use);
+            let is_delete =
+                is_delete(reference, &array_used_in_for).is_some_and(|is_assignee| is_assignee);
+
+            array_in_use && !is_delete
+        };
+
+        if references.iter().all(index_only_used_with_array) {
+            Some(())
+        } else {
+            None
+        }
+    }
+
+    fn diagnostic(node: &RuleContext<Self>, _: &Self::State) -> Option<RuleDiagnostic> {
+        Some(RuleDiagnostic::new(
+            rule_category!(),
+            node.query().range(),
+            markup! {
+                "Use "<Emphasis>"for-of"</Emphasis>" loop instead of a "<Emphasis>"for loop"</Emphasis>"."
+            },
+        ))
+    }
+}
+
+/// Get initializer references taking scope to account (excluding initials `for` loop references)
+fn get_initializer_references(
+    block: Option<&biome_js_syntax::JsBlockStatement>,
+    binding: &biome_js_syntax::JsIdentifierBinding,
+    model: &biome_js_semantic::SemanticModel,
```

Comment:
We tend to have namespaces like `biome_js_semantic::` at the top unless they are needed we have naming clash.